### PR TITLE
Add window sql support

### PR DESCRIPTION
**Pull request by joykent99**
This changelist adds window support to Druid sql module.

- DruidRules.rules() is modified to accept Window and Project on Window.
- DruidRules.DruidOuterQueryRule is modified to accept Window as subqueries.
- PartialDruidQuery is modified to accept Window and WindowProject.
- DruidQuery is modified to construct WindowBaseQuery or GroupByOverWindowBaseQuery
  when requested.
- The actual implementation of WindowBaseQuery and GroupByOverWindowBaseQuery is
  in au-druid repository.

### CODE REVIEW

**Comment by joykent99** [comment]
The plan is to use LEAD/LAG window functions to compute `duration`. The main implementation is in `au-druid`. This changelist just enables windowing in SQL module.

**Comment by jorbay-au** [comment]
talked offline about lack of tests with Joy. This can be tested in au-druid's tests, however, it would be ideal to be able to test this here before getting to au-druid. Joy suggested using a mock query window query support. I'll wait on seeing how this is implemented before accepting

**Comment by jorbay-au** [comment]
As for the code, it looks good to me, but I'm pretty out of my element with calcite, so if you know anyone with calcite experience, I recommend getting there check

**Comment by joykent99** [comment]
@jorbay-au I have added a few unittests by mocking window-related classes.

**Comment by jorbay-au** [comment]
reviewing now

File: sql/src/test/java/org/apache/druid/sql/calcite/window/WindowQueryTest.java [hunk 1]
```diff
@@ -0,0 +1,445 @@
...
+    catch (ValidationException exp) {
+      failed = true;
+      assertTrue(exp.getMessage().contains("OVER clause is necessary for window functions"));
+    }
+    assertTrue(failed);
+
+    // Different partition by clauses
+    try {
+      failed = false;
+      sql = "SELECT lag(dim1) OVER (partition by dim2), dim1, lead(dim1) OVER (), dim2, cnt from foo";
```
Comment (jorbay-au) [inline sql/src/test/java/org/apache/druid/sql/calcite/window/WindowQueryTest.java]: I didn't realize queries couldn't have clauses that partition differently. good test

File: sql/src/test/java/org/apache/druid/sql/calcite/window/WindowQueryTest.java [hunk 2]
```diff
@@ -0,0 +1,445 @@
...
+    queryContext.put(DruidQuery.WINDOW_QUERY_CONTEXT_NAME,
+                     DruidQuery.WINDOW_QUERY_TEST_CONTEXT + ":org.apache.druid.sql.calcite.window.MockWindowQueryFactory");
+
+    String sql = "SELECT lag(dim1) OVER (), dim1, lead(dim1) OVER () from foo";
+    testQuery(PLANNER_CONFIG_DEFAULT, queryContext, sql, CalciteTests.REGULAR_USER_AUTH_RESULT, 3);
+
+    sql = "SELECT lag(dim1) OVER (), dim1, lead(dim2) OVER (), dim2 from foo";
+    testQuery(PLANNER_CONFIG_DEFAULT, queryContext, sql, CalciteTests.REGULAR_USER_AUTH_RESULT, 4);
+
+    sql = "SELECT lag(dim1) OVER (partition by cnt), dim1, lead(dim2) OVER (partition by cnt), dim2, cnt from foo";
```
Comment (jorbay-au) [inline sql/src/test/java/org/apache/druid/sql/calcite/window/WindowQueryTest.java]: These look good.

File: sql/src/test/java/org/apache/druid/sql/calcite/window/WindowQueryTest.java (hunk 1 above)
Comment (joykent99) [reply sql/src/test/java/org/apache/druid/sql/calcite/window/WindowQueryTest.java]: Yes, this is the limitation of current implementation.

**Comment by joykent99** [comment]
Add @carlosgarciaalvarado to the review.

**Comment by joykent99** [comment]
Add @cramja to review.

File: processing/src/main/java/org/apache/druid/query/aggregation/AggregatorUtil.java [hunk 3]
```diff
@@ -110,6 +110,10 @@
   public static final byte HLL_SKETCH_TO_STRING_CACHE_TYPE_ID = 0x31;
   public static final byte HLL_SKETCH_TO_ESTIMATE_AND_BOUNDS_CACHE_TYPE_ID = 0x32;
 
+  // LEAD/LAG aggregator
+  public static final byte LEAD_AGG_CACHE_TYPE_ID = 0x5A;
```
Comment (carlosgarciaalvarado) [inline processing/src/main/java/org/apache/druid/query/aggregation/AggregatorUtil.java]: @joykent99 Where is this used?

Comment (joykent99) [reply processing/src/main/java/org/apache/druid/query/aggregation/AggregatorUtil.java]: This is used in the series of `LeadAggregatorFactory` (one for each data type) in `au-druid` repository -- the other code review.

Comment (joykent99) [reply processing/src/main/java/org/apache/druid/query/aggregation/AggregatorUtil.java]: It is used to define the cache key.

Comment (carlosgarciaalvarado) [reply processing/src/main/java/org/apache/druid/query/aggregation/AggregatorUtil.java]: what's the rule for the hex number? is this a unique number picked at random?

Comment (joykent99) [reply processing/src/main/java/org/apache/druid/query/aggregation/AggregatorUtil.java]: I don't know any rules for the hex number -- just randomly pick. I left a little gap here, hopefully can minimize the conflicts if we pick up more changes from `incubator-druid` later.

File: sql/src/test/java/org/apache/druid/sql/calcite/window/WindowQueryTest.java [hunk 4]
```diff
@@ -0,0 +1,445 @@
...
+    testQuery(PLANNER_CONFIG_DEFAULT, queryContext, sql, CalciteTests.REGULAR_USER_AUTH_RESULT, 5);
+  }
+
+  @Test
+  public void testWindowQuery_fail() throws Exception
```
[the remark this answers is not in the document]
Comment (joykent99) [reply sql/src/test/java/org/apache/druid/sql/calcite/window/WindowQueryTest.java]: Sure. Will do.

**Pull request merged**